### PR TITLE
Feature/user conditions

### LINKED ISSUE #22: Fix existing cypress tests

### LINKED ISSUE #27: handle isLoggedIn user flows
- set default isLoggedIn state to `false`
- If user goes to the home path, `/`, and !isLoggedIn, redirect to `/trending`
- if isLoggedIn, nav tab is visible, else hidden
- add Log In / Log Out options to user area
- if !isLoggedIn, Album view "Save" and "Add to Journal" buttons hidden

### LINKED ISSUE #28: review/add cypress tests
Make sure to add some 404 & 500 level tests
Double-check all tests are intercepted.

### PR DESCRIPTION
**Pull request by jalbe0076**
This PR will:
- close #22 
- close #27 
- close #28 

## What I did:
- Implemented Context API (sorry, I forgot about my other branch and never opened up a PR for it. So it's included in this one)
- Adjusted, stubbed and expanded tests for Trending, Collections, Search and Login - sad paths included
- Refactored network fetch requests and implemented an `handleError` function to display the issue on the DOM
- Completed Demo user buttons to populate the login info
- Implemented functionality to allow non-logged-in users to view and search albums
- Implemented functionality to navigate non-logged-in users to the `/` path if the path is manually entered in the URL

## Testing Notes/Extras: 
- `npm run cypress` to test the Trending, Collections, Search and Login tests
- Sad network and non-existing pages tests are located in the Login tests
- Fixed Search tests to test for when there are no logged-in users
- Tested login features for all types of users and checked to see what would happen after logging in
- Tested feature page and what you would see with a logged-in user and no logged-in user
- tested Collection tests for multiple users
- All adjusted tests are stubbed. 

Did this break anything?

- [X]  No
- [ ]  Yes

Type of change

- [X]  New feature (non-breaking change which adds functionality)
- [ ]  Bug fix (non-breaking change which fixes an issue)
- [X]  Refactor(DRY-ing up/ reorganizing code, etc.)
- [ ]  Super small fix (Corrected a typo, removed a comment, etc.)
- [ ]  Skip all the other stuff and briefly explain the fix.

Checklist:

- [X]  If this code needs to be tested, all tests are passing
- [X]  The code runs locally
- [ ]  There are comments where clarification/ organization is needed
- [X]  The code is DRY. If it's not, I tried my best
- [X]  I reviewed my code before pushing

What's next?
- Fix the styling on some elements. 
- Make sure all sad path tests are displayed on the dom
- Review and finalize testing
- Create GIFs to add to the readme

### CODE REVIEW

**Review by jmartin777 — COMMENTED** [review]
Really like this feature for the user/non-user interface

File: src/components/Nav/Nav.js [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
       <Link to='/' className='title' ><h1>SOUND STASH</h1></Link>
+      {currentUser && 
       <nav className='navigation-tabs'>
         <NavLink to="/collections" className='nav'>COLLECTIONS</NavLink>
         <NavLink to="/journal" className='nav'>JOURNAL</NavLink>
         <NavLink to="/discover" className='nav'>DISCOVER</NavLink>
-      </nav>
-      <p className='user-profile'>User</p>
+      </nav>}
+      {!currentUser && <button className='user-profile'>LOGIN</button>}
```
Comment (jmartin777) [inline src/components/Nav/Nav.js]: Really like this feature for the user/non-user interface navigation

File: src/components/MockData/mockusers.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 const mockusers = [
-  { userId: 1, username: 'user1', password: 'password', journal: [], collections: [] },
+  { userId: 1, username: 'user1', password: 'sound-stash', journal: [], collections: [] },
   { 
     userId: 2, 
     username: 'user2', 
-    password: 'password2', 
+    password: 'sound-stash',
```
Comment (alivaditis) [inline src/components/MockData/mockusers.js:6]: lol good call

File: src/components/Login/Login.js [hunk 3]
```diff
@@ -60,11 +60,14 @@ const Login = ({ onLogin }) => {
...
-            Login
-          </button>
+          <button className="standard-btn" type="submit">LOGIN</button>
           {loginError && <p className="error-message">{loginError}</p>}
         </form>
+        <div className='demo-user-container'>
+          <button className="standard-btn" >DEMO NO USER</button>
+          <button className="standard-btn" >DEMO USER 1</button>
+          <button className="standard-btn" >DEMO USER 2</button>
+        </div>
```
Comment (jmartin777) [inline src/components/Login/Login.js]: So are we using these buttons for showcasing the App with/without a login user-flow of typing?

File: src/components/Discover/Discover.js [hunk 4]
```diff
@@ -3,11 +3,17 @@ import React, { useState, useEffect, useContext } from "react"
...
+import { useNavigate } from 'react-router-dom';
 
 
-const Discover = ({ trendingData }) => {
-  const { currentUser } = useContext(UserContext)
+const Discover = ({ trendingData, handleApiError }) => {
+  const { currentUser, isUserLoggedIn } = useContext(UserContext)
   const [ recommendedData, setRecommendedData ] = useState()
+  const navigate = useNavigate();
+console.log(trendingData)
```
Comment (alivaditis) [inline src/components/Discover/Discover.js:13]: we can remove this console log

File: src/components/Album/Album.js [hunk 5]
```diff
@@ -20,12 +20,11 @@ function Album() {
...
       })
-      .finally(() => {
-        setLoading(false);
-      });
-  }, [id, currentUser]);
+      
+      return () => setLoading(false);
+  }, [id]);
 
   const handleAddToCollections = () => {
```
Comment (alivaditis) [inline src/components/Album/Album.js:29]: Noticing a console error when multiple albums with the same masterId are added saying key should be unique, do we want to add some logic here so they can only be added once?

**Review by alivaditis — CHANGES_REQUESTED** [review]
Really awesome job getting this feature all tied together and including sad path network request errors!

File: src/components/Login/Login.js (hunk 3 above)
Comment (jalbe0076) [reply src/components/Login/Login.js]: It's for demo purposes. you can also just manually visit the `/` path and you won't be logged in

File: src/components/Album/Album.js (hunk 5 above)
Comment (jalbe0076) [reply src/components/Album/Album.js:29]: I also noticed that from Josh's branch. I added this in the descriptions of the sad path ticket to just get things moving.

**Review by alivaditis — APPROVED** [review]

**Review by jmartin777 — APPROVED** [review]
Great job Jason !

**Pull request merged**